R/compound_taskid-utils.R

```diff
@@ -0,0 +1,203 @@
+#' Detect the compound_taskid_set for a tbl for each modeling task in a given round.
+#'
+#' @param tbl a tibble/data.frame of the contents of the file being validated.
+#' Column types must **all be character**.
+#' @param config_tasks a list representantion of the `tasks.json` config file.
+#' @param round_id Character string. The round ID.
+#' @param compact Logical. If TRUE, the output will be compacted to remove NULL elements.
+#' @param error Logical. If TRUE, an error will be thrown if the compound task ID set is not valid.
+#' If FALSE and an error is detected, the detected compound task ID set will be
+#' returned with error attributes attached.
+#'
+#' @return A list of vectors of compound task IDs detected in the tbl, one for each
+#' modeling task in the round. If `compact` is TRUE, modeling tasks returning NULL
+#' elements will be removed.
+#' @noRd
+get_tbl_compound_taskid_set <- function(tbl, config_tasks, round_id,
+                                        compact = TRUE, error = TRUE) {
+  tbl <- tbl[tbl$output_type == "sample", names(tbl) != "value"]
+  out_tid <- hubUtils::std_colnames["output_type_id"]
+
+  mt_compound_taskids <- get_round_compound_task_ids(
+    config_tasks,
+    round_id
+  )
+
+  mt_tbls <- purrr::map(
+    .x = expand_model_out_grid(
+      config_tasks = config_tasks,
+      round_id = round_id,
+      all_character = TRUE,
+      include_sample_ids = FALSE,
+      bind_model_tasks = FALSE
+    ),
+    function(.x) {
+      if (!has_spls_tbl(.x)) {
+        return(NULL)
+      }
+      dplyr::inner_join(tbl, .x[, names(.x) != out_tid],
+        by = setdiff(names(tbl), out_tid)
+      )
+    }
+  )
+
+  call <- rlang::current_env()
+  tbl_compound_taskids <- purrr::map2(
+    mt_tbls, mt_compound_taskids, function(.x, .y) {
+      get_mt_compound_taskid_set(.x, .y, config_tasks,
+        error = error, call = call
+      )
+    }
+  ) %>%
+    purrr::set_names(seq_along(.))
+
+  if (compact) {
+    tbl_compound_taskids <- purrr::compact(tbl_compound_taskids)
+  }
+
+  tbl_compound_taskids
+}
+
+# Detect the compound_taskid_set for a tbl for a single modeling task.
+get_mt_compound_taskid_set <- function(tbl, config_comp_tids, config_tasks,
+                                       error = TRUE, call = NULL) {
+  if (any(is.null(tbl), is.null(config_comp_tids))) {
+    return(NULL)
+  }
+
+  if (nrow(tbl) == 0L) {
+    cli::cli_alert_info("No samples found in the file.")
+    return(NULL)
+  }
+  if (is.null(call)) call <- rlang::current_env()
+  out_tid <- hubUtils::std_colnames["output_type_id"]
+  task_ids <- hubUtils::get_task_id_names(config_tasks)
+
```

Comment:
For extra context on the following code see https://github.com/hubverse-org/hubValidations/issues/88#issuecomment-2222467279